[hop][inductor] track the dependency on unbacked symbols correctly with constant_args for hops

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143456

Before the PR, we're getting an undefined symbol error for output code when an unbacked symint is **only** used in the hop because we didn't correctly record the dependency of the unbacked symbols for hops and it gets DCEed accidentally. 

This PR adds the symbol arguments to `constant_args`, where the dependencies can be correctly constructed when `get_unbacked_symbol_uses` is called to check constant_args. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

ydwu4 committed:
[hop] support unbacked symint closure in inductor

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop] support unbacked symint closure in inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop] support unbacked symint closure in inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop] support unbacked symint closure in inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop] support unbacked symint closure in inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

zou3519 commented on torch/_inductor/ir.py:
```diff
@@ -5340,9 +5340,18 @@ def canonicalize(self):  # type: ignore[no-untyped-def]
-        # NB: It's not necessary to check regular inputs as we automatically
-        # have dependencies on them
         r: OrderedSet[sympy.Symbol] = OrderedSet()
+        # NB: It's not necessary to check regular tensor inputs as we automatically
```
NB: We're doing this because... there was previously an assumption that an unbacked symint has an associated Tensor. However this is not the case in while_loop -- we trace int inputs as unbacked symints.

ydwu4 committed:
Update on "[hop] support unbacked symint closure in inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop] support unbacked symint closure in inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop] support unbacked symint closure in inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop] support unbacked symint closure in inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop] support unbacked symint closure in inductor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop][inductor] track the dependency on unbacked symbols correctly with constant_args for hops"


Before the PR, we're getting an undefined symbol error for output code when an unbacked symint is **only** used in the hop because we didn't correctly record the dependency of the unbacked symbols for hops and it gets DCEed accidentally. 

This PR adds the symbol arguments to `constant_args`, where the dependencies can be correctly constructed when `get_unbacked_symbol_uses` is called to check constant_args. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

desertfire commented on torch/_inductor/ir.py:
```diff
@@ -7210,15 +7212,15 @@ def codegen(self, wrapper) -> None:  # type: ignore[no-untyped-def]
 class Conditional(ExternKernel):
     predicate: Optional[IRNode] = None
-    operands: Optional[list[TensorBox]] = None
+    operands: Optional[List[Union[TensorBox, ShapeAsConstantBuffer]]] = None
```
List->list is a recent update, https://github.com/pytorch/pytorch/pull/145137. Please don't undo it.

desertfire commented on torch/_inductor/ir.py:
```diff
@@ -7248,11 +7248,9 @@ def create(  # type: ignore[no-untyped-def]
-        operands: list[TensorBox],
+        operands: List[Union[TensorBox, ShapeAsConstantBuffer]],
     ):
-        predicate = cls.realize_input(predicate)
```
No need to realize predicate now?

ydwu4 replied:
pred is either a scalar tensor or a sym expression. There's not much fusion opportunities left for downstream operators. Realizing it also reduces the types that we need to handle later on.

ydwu4 replied:
Fix the typo

eellison reviewed (commented):
didn't look too closely bc failing tests, re-review when ready

ydwu4 committed:
Update on "[hop][inductor] track the dependency on unbacked symbols correctly with constant_args for hops"


Before the PR, we're getting an undefined symbol error for output code when an unbacked symint is **only** used in the hop because we didn't correctly record the dependency of the unbacked symbols for hops and it gets DCEed accidentally. 

This PR adds the symbol arguments to `constant_args`, where the dependencies can be correctly constructed when `get_unbacked_symbol_uses` is called to check constant_args. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop][inductor] track the dependency on unbacked symbols correctly with constant_args for hops"


Before the PR, we're getting an undefined symbol error for output code when an unbacked symint is **only** used in the hop because we didn't correctly record the dependency of the unbacked symbols for hops and it gets DCEed accidentally. 

This PR adds the symbol arguments to `constant_args`, where the dependencies can be correctly constructed when `get_unbacked_symbol_uses` is called to check constant_args. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop][inductor] track the dependency on unbacked symbols correctly with constant_args for hops"


Before the PR, we're getting an undefined symbol error for output code when an unbacked symint is **only** used in the hop because we didn't correctly record the dependency of the unbacked symbols for hops and it gets DCEed accidentally. 

This PR adds the symbol arguments to `constant_args`, where the dependencies can be correctly constructed when `get_unbacked_symbol_uses` is called to check constant_args. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 committed:
Update on "[hop][inductor] track the dependency on unbacked symbols correctly with constant_args for hops"


Before the PR, we're getting an undefined symbol error for output code when an unbacked symint is **only** used in the hop because we didn't correctly record the dependency of the unbacked symbols for hops and it gets DCEed accidentally. 

This PR adds the symbol arguments to `constant_args`, where the dependencies can be correctly constructed when `get_unbacked_symbol_uses` is called to check constant_args. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

desertfire commented on torch/_inductor/ir.py:
```diff
@@ -20,6 +20,7 @@
     Literal,
     Optional,
     overload,
+    Tuple,
```
Er, a similar story here, Tuple->tuple

desertfire reviewed (approved):
LGTM other than the tuple annotation.

ydwu4 committed:
Update on "[hop][inductor] track the dependency on unbacked symbols correctly with constant_args for hops"


Before the PR, we're getting an undefined symbol error for output code when an unbacked symint is **only** used in the hop because we didn't correctly record the dependency of the unbacked symbols for hops and it gets DCEed accidentally. 

This PR adds the symbol arguments to `constant_args`, where the dependencies can be correctly constructed when `get_unbacked_symbol_uses` is called to check constant_args. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "[hop][inductor] track the dependency on unbacked symbols correctly with constant_args for hops"


Before the PR, we're getting an undefined symbol error for output code when an unbacked symint is **only** used in the hop because we didn't correctly record the dependency of the unbacked symbols for hops and it gets DCEed accidentally. 

This PR adds the symbol arguments to `constant_args`, where the dependencies can be correctly constructed when `get_unbacked_symbol_uses` is called to check constant_args. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

atalman commented:
@pytorchmergebot revert -c nosignal -m "New tests are failing internally"

ydwu4 committed:
Update on "[hop][inductor] track the dependency on unbacked symbols correctly with constant_args for hops"


Before the PR, we're getting an undefined symbol error for output code when an unbacked symint is **only** used in the hop because we didn't correctly record the dependency of the unbacked symbols for hops and it gets DCEed accidentally. 

This PR adds the symbol arguments to `constant_args`, where the dependencies can be correctly constructed when `get_unbacked_symbol_uses` is called to check constant_args. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
